### ISSUE TITLE
Inconsistent scoping rules when using scan with a break condition

### ISSUE THREAD

**Issue by cknd** [Theano/Theano#4655]
_**tl;dr:** `scan` already knows how to infer missing input variables for any **output expression** of its inner function. I found out the hard way that it does not attempt to discover inputs of the expression that defines the **break condition**._

I'm on theano 0.9.0dev1.

My situation:
I have a while-loop with a very complicated break condition. It's an expression whose input variables live far away in the code base and which can easily change. So it would be cumbersome to provide all of the inputs for the break condition using scan's `non_sequences` argument. Luckily, the computational graph of the break expression already contains this knowledge! But it seems that scan currently does not use it.

This means there are two different rules for variable scoping when you write an inner function for a scan op: Input variables that feed into the scan-fn's outputs are discovered, variables that feed into its break expression are not.

Below is a minimal example, some rough idea of what piece of scan's code might be involved and a user-side workaround.

(of course the example is simple enough that I could just use `non_sequences`, but let's pretend that that would be very horrible.)

```python
import theano
from theano import tensor as T
print(theano.__version__)
```

    0.9.0dev1.dev-RELEASE

```python

X = T.scalar('x')

# this inner function works fine. *thumbs up*
# (`scan` infers that `X` is a required input for the output expression `val`)
def fn1(val):
    val = val*X
    until = theano.scan_module.until(val <= 0.01)
    return val, until
result, _ = theano.scan(fn=fn1, outputs_info=[1.], non_sequences=[], n_steps=4)
print(result.eval({X:0.1}))
```

    [ 0.1    0.01   0.001]

```python

X = T.scalar('x')

# This inner function throws a MissingInputError, asking for X.
# The only difference is that X now isn't part of the computation 
# graph of any of the output variables any more. It's only in the 
# expression that defines the stop condition.
def fn2(val):
    val = val*0.1
    some_complex_thing = val*X   # or whatever. imagine something with dozens of outside variables.
    until = theano.scan_module.until(some_complex_thing <= 0.01)
    return val, until

result, _ = theano.scan(fn=fn2, outputs_info=[1.], non_sequences=[], n_steps=4)
print(result.eval({X:0.01}))
```

    MissingInputError                         Traceback (most recent call last)
    (....)

This exception pops up in `theano.scan` [at the point where the dummy function is compiled.](https://github.com/Theano/Theano/blob/7534881b5b5239570690596a344e19505fae5d32/theano/scan_module/scan.py#L806-L832)
It looks like input arguments that haven't been supplied via `non_sequences` are auto-completed by a call to `gof.graph.inputs`, but this is only done [on the output variables of the scan function](https://github.com/Theano/Theano/blob/7534881b5b5239570690596a344e19505fae5d32/theano/scan_module/scan.py#L809), not on the expression that defines the break condition. (still the conditional expression is appended to `dummy_outs`, which explains why the dummy function won't compile.)

Anyway. Workaround:

**I really wanted to rely on scan's automatic inference of parent variables** instead of supplying all of them via non_sequences. So to get convenient automatic scoping, I to made the condition expression part of the output list (and then later I ignore its result value).

```python
# workaround
# workaround
def fn3(val, some_complex_thing):
    val = val*0.1
    some_complex_thing = val*X  # or whatever
    until = theano.scan_module.until(some_complex_thing <= 0.01)
    return [val, some_complex_thing], until

some_complex_thing_init = 0.
result, _ = theano.scan(fn=fn3, outputs_info=[1., some_complex_thing_init], non_sequences=[], n_steps=4)
print(result[0].eval({X:0.1}))
```

    [ 0.1     0.01]
